client/src/test/java/hudson/plugins/swarm/YamlConfigTest.java

```diff
@@ -0,0 +1,209 @@
+
+        final Options linear = loadYaml("url: ignore\nretryBackOffStrategy: LINEAR\n");
+        assertThat(linear.retryBackOffStrategy, equalTo(RetryBackOffStrategy.LINEAR));
+
+        final Options exponential = loadYaml("url: ignore\nretryBackOffStrategy: EXPONENTIAL\n");
+        assertThat(exponential.retryBackOffStrategy, equalTo(RetryBackOffStrategy.EXPONENTIAL));
+    }
+
+    @Test
+    public void modeOptionValues() throws ConfigurationException {
+        final Options normal = loadYaml("url: ignore\nmode: " + ModeOptionHandler.NORMAL + "\n");
+        assertThat(normal.mode, equalTo(ModeOptionHandler.NORMAL));
+        final Options exclusive =
+                loadYaml("url: ignore\nmode: " + ModeOptionHandler.EXCLUSIVE + "\n");
+        assertThat(exclusive.mode, equalTo(ModeOptionHandler.EXCLUSIVE));
+    }
+
+    @Test
+    public void failIfModeHasInvalidValue() {
+        final Throwable ex =
+                assertThrows(
+                        ConfigurationException.class,
+                        () -> loadYaml("url: ignore\nmode: invalid\n"));
+        assertThat(ex.getMessage(), containsString("mode"));
+    }
+
+    @Test
+    public void failsIfDeprecatedOptionIsUsed() {
+        final Throwable ex =
+                assertThrows(
+                        ConfigurationException.class, () -> loadYaml("password: should-fail\n"));
+        assertThat(ex.getMessage(), containsString("password"));
+    }
+
+    @Test
+    public void failsOnWebSocketAndTunnelOption() {
+        final Throwable ex =
+                assertThrows(
+                        ConfigurationException.class,
+                        () ->
+                                loadYaml(
+                                        "url: ignore\n"
+                                                + "webSocket: true\n"
+                                                + "tunnel: excluded-by.ws:123\n"));
+        assertThat(ex.getMessage(), containsString("tunnel"));
+    }
+
+    @Test
+    public void failsOnDisableWorkDirAndWorkDirOption() {
+        final Throwable ex =
+                assertThrows(
+                        ConfigurationException.class,
+                        () -> loadYaml("url: ignore\ndisableWorkDir: true\nworkDir: /tmp/wd\n"));
+        assertThat(ex.getMessage(), containsString("workDir"));
+    }
+
+    @Test
+    public void failsOnDisableWorkDirAndInternalDirOption() {
+        final Throwable ex =
+                assertThrows(
+                        ConfigurationException.class,
+                        () ->
+                                loadYaml(
+                                        "url: ignore\n"
+                                                + "disableWorkDir: true\n"
+                                                + "internalDir: /tmp/id\n"));
+        assertThat(ex.getMessage(), containsString("internalDir"));
+    }
+
+    @Test
+    public void failsOnDisableWorkDirAndFailIfWorkDirIsMissingOption() {
```

Comment:
Some of these tests seem redundant now that we've made the validation code generic. It isn't necessary to write a test case for each possible conflict. A single test case that exercises the conflict handling code is sufficient.

Reply:
Done. A good chance to add testing for both options names too.